BootloaderCorePkg/Library/SmbiosInitLib/SmbiosInitLib.c

```diff
@@ -0,0 +1,389 @@
+  )
+{
+  SMBIOS_TABLE_TYPE127          *Type127;
+  EFI_STATUS                    Status;
+  SMBIOS_TABLE_ENTRY_POINT      *SmbiosEntry;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if (SmbiosEntry == NULL) {
+    return;
+  }
+
+  //
+  // Create Type127 and update global Type127 for the next append
+  //
+  Type127 = (SMBIOS_TABLE_TYPE127 *) (SmbiosEntry->TableAddress + SmbiosEntry->TableLength);
+  Type127->Hdr.Type = SMBIOS_TYPE_END_OF_TABLE;
+  Type127->Hdr.Length = sizeof (SMBIOS_STRUCTURE);  
+  mType127Ptr = (VOID *) Type127;
+
+  //
+  // Update the Table Length and check for overflows
+  //
+  SmbiosEntry->TableLength += Type127->Hdr.Length + 2;
+  Status = CheckSmbiosOverflow ();
+  if (EFI_ERROR(Status)) {
+    DEBUG ((DEBUG_ERROR, "Smbios Tables Overflown!"));
+    return;
+  }
+
+  //
+  // Patch entry point struct with new data
+  //
+  if ((UINT16)NewMaxStructSize > SmbiosEntry->MaxStructureSize) {
+    SmbiosEntry->MaxStructureSize = (UINT16)NewMaxStructSize;
+  }
+  SmbiosEntry->IntermediateChecksum = SmbiosEntry->EntryPointStructureChecksum = 0;
+  SmbiosEntry->IntermediateChecksum         = CalculateCheckSum8 ((UINT8 *)SmbiosEntry + 0x10, sizeof (SMBIOS_TABLE_ENTRY_POINT) - 0x10);
+  SmbiosEntry->EntryPointStructureChecksum  = CalculateCheckSum8 ((UINT8 *)SmbiosEntry       , sizeof (SMBIOS_TABLE_ENTRY_POINT)       );
+}
+
+/**
+  Append an Smbios Type to an existing set of types
+  Different from AddSmbiosType, which gets called during initial Smbios setup
+
+  @param[in]  TypeData    pointer to the Type Data including strings
+  @param[in]  TypeLength  Size of type data including strings and end terminator (0000)
+
+**/
+VOID
+AppendSmbiosType (
+  IN  VOID      *TypeData,
+  IN  UINTN     TypeLength
+  )
+{
+  SMBIOS_TABLE_ENTRY_POINT    *SmbiosEntry;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if (SmbiosEntry == NULL || mType127Ptr == NULL) {
+    return ;
+  }
+
+  //
+  // Find current Type127 and replace it with Type Data
+  //
+  SmbiosEntry->TableLength -= ( ((SMBIOS_STRUCTURE *)mType127Ptr)->Length + 2);
+  CopyMem (mType127Ptr, TypeData, TypeLength);
+  SmbiosEntry->TableLength += (UINT16)TypeLength;
+
+  //
+  // After appending, update with Typ127 and patch entry point
+  //
+  FinalizeSmbios (TypeLength);
+}
+
+/**
+  Get the value a particular string in a Type
+  from data structure pointed to by PcdSmbiosStringsPtr
+
+  @param[in]  Type      Get the string for a Type
+  @param[in]  StringId  String Num
+
+  @retval               CHAR8 * to the string, if found
+                        "Unknown"            , otherwise
+
+**/
+CHAR8 *
+GetSmbiosString (
+  IN  UINT8   Type,
+  IN  UINT8   StringId
+  )
+{
+  SMBIOS_TYPE_STRINGS      *SmbiosStrings;
+
+  SmbiosStrings = (SMBIOS_TYPE_STRINGS *) PcdGet32 (PcdSmbiosStringsPtr);
+
+  if (SmbiosStrings == NULL) {
+    return "Unknown";
+  }
+
+  for (; SmbiosStrings->Type != 127;) {
```

Comment:
Can we limit the max loop count ?    EX: If there is no type 127 in string table,  the code will access data outside of the string table.